src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -407,23 +407,55 @@ private static object BindToCollection(Type type, IConfiguration config, BinderO
             }
         }
 
-        [RequiresUnreferencedCode("In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
-        private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type type)
+        [RequiresUnreferencedCode(
+            "In case type is a Nullable<T>, cannot statically analyze what the underlying type is so its members may be trimmed.")]
+        private static object CreateInstance(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors |
+                                        DynamicallyAccessedMemberTypes.NonPublicConstructors)]
+            Type type,
+            IConfiguration config,
+            BinderOptions options)
         {
             Debug.Assert(!type.IsArray);
 
-            if (type.IsAbstract)
+            if (type.IsInterface || type.IsAbstract)
             {
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(BindingFlags.Public | BindingFlags.Instance);
+
+            bool hasParameterlessConstructor =
+                constructors.Any(ctor => ctor.GetParameters().Length == 0 || type.IsValueType);
+
+            if (!type.IsValueType && constructors.Length == 0)
+            {
+                throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
```

Comment:
I wonder if this exception message should be updated now. Before it had to be a parameterless constructor. But now, we accept other kinds of constructors as well.

Reply:
Good spot. Updated.